diffpy/snmf/subroutines.py

```diff
@@ -49,3 +50,35 @@ def objective_function(residual_matrix, stretching_factor_matrix, smoothness, sm
     component_matrix = np.asarray(component_matrix)
     return .5 * np.linalg.norm(residual_matrix, 'fro') ** 2 + .5 * smoothness * np.linalg.norm(
         smoothness_term @ stretching_factor_matrix.T, 'fro') ** 2 + sparsity * np.sum(np.sqrt(component_matrix))
+
+
+def get_stretched_component(stretching_factor, component, signal_length):
+    """Applies a stretching factor to a component signal.
+
+    Approximates the values of a component signal at points in between its grid nodes using quadratic spline
+    interpolation. Uses a normalized grid of evenly spaced integers counting from 0 to signal_length (exclusive) to
+    approximate values in between grid nodes. Once this grid is stretched, values at grid nodes past the unstretched
+    signal's domain are set to zero. Returns the approximate values of x(r/a) from x(r) where x is a component signal.
+
+    Parameters
+    ----------
+    stretching_factor: float
+      The stretching factor of a component signal at a particular moment.
+    component: 1d array like
+      The calculated component signal without stretching or weighting. Has length N, the length of the signal.
+    signal_length: int
+      The length of the component signal.
+
+    Returns
+    -------
+    1d array of floats
+      The calculated component signal with stretching factors applied. Has length N, the length of the unstretched
+      component signal.
+
+    """
+    component = np.asarray(component)
+    normalized_grid = np.arange(signal_length)
+    spline = scipy.interpolate.CubicSpline(normalized_grid, component)
```

Comment:
Can you make it work using numpy or scipy.interpolate functions?  It will be cleaner.  You can choose linear or cubic interpolations there.

Reply:
Do you mean having a function that does the interpolation instead of using a class like I did?

Reply:
using this instead of coding it yourself:
https://docs.scipy.org/doc/scipy/reference/interpolate.html